### PR TITLE
Expand technical guidance (secret storage, pre-commit on-device scanning for secrets)

### PR DESCRIPTION
**Pull request by joelgsamuel**
- Setting the pattern for secrets storage within our applications - KMS, Kubernetes Secrets.

- Removing the suggestion that plain text secrets in private Github.com repos are OK

- Requiring the use of git-crypt where secrets are can only be stored in code

- Establishing the use of git-secrets as a pre-commit hook to avoid secrets ever leaving the device as a result of using git

### CODE REVIEW

File: standards/storing-secrets.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+All our secrets should be adequately protected and suitably stored.
+
+Where possible, use infrastructure-based secrets management services such as [AWS Key Management Service](https://aws.amazon.com/kms/), [AWS Systems Manager Parameter Store](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html), [Microsoft Azure Key Vault](https://azure.microsoft.com/en-gb/services/key-vault/) or [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) on MOJ's Cloud Platforms.
+
+## git-crypt
+
+If you need to store secrets within Github.com, you must use [git-crypt](https://github.com/AGWA/git-crypt) to encrypt those secrets and control who has the ability to view (decrypt) those secrets.
+
+You must never store secrets or otherwise sensitive information in code repositories.
```
[the remark this answers is not in the document]
Comment (joelsamuelmoj) [reply standards/storing-secrets.md]: Fixed

File: guides/using-git-secrets.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+Installation instructions can be found in the [git-crypt repository](https://github.com/awslabs/git-secrets).
+
+### Use our base configuration
+
+We need to put in something here about where our base configuration is, meaning we likely have to create one or 'bless' the base one provided.
+
+### Scanning for the first time
+
+We should write some instructions about installing and scanning for the first time.
```
[the remark this answers is not in the document]
Comment (joelsamuelmoj) [reply guides/using-git-secrets.md]: (buck passing: I'm not going to do this)

A guide makes sense but then assumes everyone has to follow it to get going. I think there should be a very basic config they can just pull from this repo (etc)

File: guides/using-git-secrets.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+---
+# Using git-secrets
+
+## Overview
+
+We use [git][git] is a [distributed version-control system][dvcs]. Git primarily works by comparing a local working copy of repository content against another branch, highlighting and controlling any changes.
+
+As Git utilises a local folder structure (a clone of a repository) on your device, it is possible to inadvertently upload unintended files to Github.com - for example, private keys and other secrets which should not be uploaded or shared with anyone else.
+
+You must use [git-secrets](https://github.com/awslabs/git-secrets) on your device to limit the possibility of accidentally uploading secrets files or information.
```
[the remark this answers is not in the document]
Comment (joelsamuelmoj) [reply guides/using-git-secrets.md]: I think it would be a different standard (secret scanners being pre-commit pre-storage) but yes that makes sense.

I'll draft a thing.

File: guides/using-git-secrets.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+## How-to
+
+### Install git-crypt
+
+Installation instructions can be found in the [git-crypt repository](https://github.com/awslabs/git-secrets).
```
[the remark this answers is not in the document]
Comment (joelsamuelmoj) [reply guides/using-git-secrets.md]: Corrected (pitfalls of working on different docs at the same time)

File: guides/using-git-secrets.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+---
+category: Tools
+expires: 2019-01-31
+---
+# Using git-secrets
```
Comment (SteveMarshall) [inline guides/using-git-secrets.md]: I feel like this doesn't actually add any more value than what's in the standard. Can we kill this till we have a base config?

File: standards/storing-secrets.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+---
+category: Building software
+expires: 2019-01-01
+---
+# Storing secrets
+
+All our secrets should be adequately protected and suitably stored.
+
+Where possible, use infrastructure-based secrets management services such as [AWS Key Management Service](https://aws.amazon.com/kms/), [AWS Systems Manager Parameter Store](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-paramstore.html), [Microsoft Azure Key Vault](https://azure.microsoft.com/en-gb/services/key-vault/) or [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) on MOJ's Cloud Platforms.
```
Comment (SteveMarshall) [inline standards/storing-secrets.md]: Maybe link to the Cloud Platform docs for secrets (`https://ministryofjustice.github.io/cloud-platform-user-docs/02-deploying-an-app/003-add-secrets-to-deployment/`) instead of generic k8s docs?

I might also suggest rewriting this to be more directive, like:

```markdown

Use infrastructure-based secrets management services where possible.

- On the MOJ Cloud Platform, use Kubernetes secrets
- On AWS, use KMS
- On Azure, use Key Vault
```

File: standards/storing-secrets.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+## git-crypt
+
+If you do need to store secrets within Github.com they should be adequately protected.
```
Comment (SteveMarshall) [inline standards/storing-secrets.md]: “within Git repositories”

File: standards/storing-secrets.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+[git-crypt](https://github.com/AGWA/git-crypt) has previously been used by MOJDT teams to encrypt those secrets and control who has the ability to view (decrypt) those secrets.
+
+You must never store unencrypted secrets in code repositories.
+
+## Preventing secret leaks
+
+We use [git][git] is a [distributed version-control system][dvcs]. Git primarily works by comparing a local working copy of repository content against another branch, highlighting and controlling any changes.
```
Comment (SteveMarshall) [inline standards/storing-secrets.md]: Why is this here? You've mentioned git without explaining what it is already in this document? Also, the links are broken.

File: standards/storing-source-code.md [hunk 9]
```diff
@@ -21,19 +21,29 @@ New repositories for products and services live in the
...
 Repositories should be [clearly named]({{ '/standards/naming-things' | relative_url }}),
 and have an [appropriate licence]({{ '/standards/licencing-software-or-code' | relative_url }})
 and enough documentation that someone new can get started with the
 project.
 
-If the repository contains secrets rather than code, make it private
-and restrict the people who have access to it. Discuss with the
-security engineering team whether you should encrypt the contents of
-the repository.
+### GitHub.com accounts
```
Comment (SteveMarshall) [inline standards/storing-source-code.md]: `GitHub`, not `GitHub.com`; GitHub Enterprise has never been a thing at MOJ, so we don't need the qualifier.

File: standards/storing-source-code.md [hunk 10]
```diff
@@ -21,19 +21,29 @@ New repositories for products and services live in the
...
+
+You can use your personal GitHub account (but you should [add your MOJ
+email address to your account](https://help.github.com/articles/adding-an-email-address-to-your-github-account/),
+and maybe [use that for notifications](https://help.github.com/articles/managing-notification-emails-for-organizations/)).
+
+You should update your GitHub account profile so it is clear who you are, such as your display name or profile picture. This helps others easily identify who has made requests, added comments or provided approvals.
```
Comment (SteveMarshall) [inline standards/storing-source-code.md]: This isn't necessary, and requires people to increase their public information, which some people may be uncomfortable with. I'm in the process of trying to upgrade us to [GitHub Business Cloud](https://github.com/business), which will allow us to do org membership via Okta, so we won't need any of this.

File: standards/storing-source-code.md [hunk 11]
```diff
@@ -43,3 +43,7 @@ Ask the [Digital Service Desk](https://mojdt.slack.com/messages/C59CX1RHN) to ad
 We [make source code open](https://www.gov.uk/service-manual/technology/making-source-code-open-and-reusable) but there may be times where making a repository private may be the best thing to do. Using private repositories should be relatively rare.
 
 Currently due to Github.com permission structures, private repositories in the `ministryofjustice` Github.com organisation can be viewed by any member of the `ministryofjustice` organisation therefore are better described as "non-public" repositories.
+
+### Use `.gitignore`
+
+To help prevent the inadvertent upload of files not usually associated with application, configuration or infrastructure code (for example, .xlsx .tar.gz .zip .exe) use `.gitignore` so that `git` on your device will ignore these files and not upload them during a commit.
```
[the remark this answers is not in the document]
Comment (joelgsamuel) [reply standards/storing-source-code.md]: I'm happy to relax the language as whole (even more beyond "not usually associated") into "consider" (etc) but I fail to see why those files should end up in github.com. If we're "publishing", we would do so formally on GOV.UK, MOJ intranet etc.

This is guidance as well, if the person has a clear justifiable need, they can of course edit their `.gitignore` as they see fit.

Good commit hygiene and secret scanners are demonstrably unable to catch these file exclusion issues.

File: guides/using-git-secrets.md (hunk 5 above)
Comment (joelgsamuel) [reply guides/using-git-secrets.md]: Agreed, I'll leave this open as I'm going to have the base config written (this is the motivator to do it now) and then we'll have more content (and more to review) @sid234 I'll be in touch :-)

File: standards/storing-secrets.md (hunk 6 above)
Comment (joelgsamuel) [reply standards/storing-secrets.md]: Changed

File: standards/storing-secrets.md (hunk 7 above)
Comment (joelgsamuel) [reply standards/storing-secrets.md]: Changed

File: standards/storing-source-code.md (hunk 9 above)
Comment (joelgsamuel) [reply standards/storing-source-code.md]: This is referring to accounts on a website, so saying it is github.com makes sense? I appreciate if you say "github" most people think of "github.com" but maybe I'm splitting hairs to be explicit

Comment (joelgsamuel) [reply standards/storing-source-code.md]: I appreciate the change to Okta (probability of coming through and when?) but I think we should always ask people to have MFA on their github.com account anyway?

File: standards/storing-source-code.md (hunk 10 above)
Comment (joelgsamuel) [reply standards/storing-source-code.md]: As per other comment, fair enough if Okta will entirely satisfy this purpose

File: standards/storing-source-code.md (hunk 9 above)
Comment (SteveMarshall) [reply standards/storing-source-code.md]: We should definitely ask them to have MFA, yeah.

`GitHub.com` just sounds weird, is all.

File: guides/using-git-secrets.md (hunk 5 above)
Comment (SteveMarshall) [reply guides/using-git-secrets.md]: I'd rather we don't push visibly-incomplete things to master, personally. Things that need *some* attention are fair enough, but this should just be in a separate branch.

File: standards/storing-source-code.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (SteveMarshall) [reply standards/storing-source-code.md]: There are entirely valid cases where you'd want to commit those types of files. You might, for example, be writing a library that tests importing data from a zipped Excel template, and want to have test data in the repository. I can think of at least five teams to whom that exact use case might apply.

Further, unless we can specify a precise list of the entire set of files that people should include in their `.gitignore`, then I don't see why anyone should independently try to come up with a list of files they think might not be appropriate. If there are real risks associated with a given project, that project team might choose to do this, but as it's written right now, this guidance expects too much of the reader.

**Comment by joelgsamuel** [comment]
`security-guidance` has been created since this PR was submitted. Closing to avoid dual coverage.
